paasta_tools/kubernetes/bin/paasta_secrets_sync.py

```diff
@@ -396,6 +418,79 @@ def sync_secrets(
     return True
 
 
+def sync_database_credentials(
+    kube_client: KubeClient,
+    cluster: str,
+    service: str,
+    secret_provider_name: str,
+    vault_cluster_config: Dict[str, str],
+    soa_dir: str,
+    vault_token_file: str,
+) -> bool:
+    config_loader = PaastaServiceConfigLoader(service=service, soa_dir=soa_dir)
+    for instance_config in config_loader.instance_configs(
+        cluster=cluster, instance_type_class=KubernetesDeploymentConfig
+    ):
+        # expects VAULT_ADDR_OVERRIDE, VAULT_CA_OVERRIDE, and VAULT_TOKEN_OVERRIDE to be set
+        # in order to use a custom vault shard
+        database_credentials = instance_config.get_database_credentials()
+
+        provider = get_secret_provider(
+            secret_provider_name=secret_provider_name,
+            soa_dir=soa_dir,
+            service_name=service,
+            cluster_names=[cluster],
+            # overridden by env variables but still needed here for spec validation
+            secret_provider_kwargs={
+                "vault_cluster_config": vault_cluster_config,
+                "vault_auth_method": "token",
+                "vault_token_file": vault_token_file,
+            },
+        )
+
+        for dstore in database_credentials.keys():
+            for credential in database_credentials[dstore]:
+                vault_path = f"secrets/database/{dstore}/{credential}"
+                secrets = provider.get_secrets_from_vault_path(vault_path)
+                if not secrets:
+                    # no secrets found at this path. skip syncing
+                    log.debug(
+                        f"Warning: no secrets found at requested path {vault_path}."
+                    )
+                    continue
+
+                # decrypt and save in secret_data
+                vault_key_path = get_vault_key_secret_name(vault_path)
+
+                # kubernetes expects data to be base64 encoded binary in utf-8 when put into secret maps
+                secret_data = {}
+                secret_data[vault_key_path] = base64.b64encode(
+                    json.dumps(secrets).encode("utf-8")
```

Comment:
might be good to have a comment for what `secrets` looks like

additionally, is there any way that this is something that's not json-serializable?

Reply:
++ will add a comment

for our use case, it'll always be JSON serializable, though it wouldn't hurt to add a check if you suggest